Add exclude option for penthouse `forceExclude` config

batbattur opened the pull request:
## Summary
Added option to exclude certain css selectors from the critical css using the `forceExclude` option in penthouse.

## QA notes
We can run the `critical-css.yml` workflow from the `iFixit/ifixit` side (with this branch specified) to verify it still works. 
qa_req 0

connects [#47316](https://github.com/ifixit/ifixit/issues/47316)

batbattur committed:
Add option to pass exclude argument for critical-css.js file

The --exclude option will be used to set the `forceExclude` option
in penthouse config. `forceExclude` expects an array of selectors or
an empty array.

Later on we will be calling the critical-css.js using a command like
the following: node critical-css.js URL --exclude=.selector1 .selector2.

batbattur committed:
Add option to pass --exclude option to run.rb script

In reduce_to_critical, we pass the exclude arguments to the
critical-css.js script, therefore we are adding the --exclude
option inside the run.rb script.

batbattur committed:
Fix when passing regex using --exclude option

Before, it was treating the regex as a string so it didn't work.
Now it checks if the exclude argument is a regex.

batbattur committed:
Add --exclude example in readme

batbattur commented:
QA 🚛 tested the changes in https://github.com/iFixit/ifixit/pull/47367#issuecomment-1505796048

ardelato commented on critical-css.js:
```diff
@@ -34,6 +34,13 @@ function main(urls) {
+  // When a regex is passed from a command line it converts to a string,
+  // so we need to convert it back to a regex
+  const excludeArr = args['exclude'].map((arg) =>
+    arg.startsWith("/") && arg.endsWith("/")
```
Would this work if we pass regex flags? ie `/\s/g` or `/\w/gi`

batbattur replied:
Thanks for catching this. I think it will fail to catch if the ending is not `/`. I'll look for a different way to check if it's a regex.

batbattur replied:
Fixed in https://github.com/iFixit/css-gather/pull/9/commits/89f826b2ee03e83ac613e1d0e8d9cd24c9624e47.

ardelato reviewed (approved):
CR 👍 but deploy_block ☁️ on question

batbattur commented:
dev_block 🚛

batbattur committed:
Check regex in a string properly

Created a new function to check if the string contains a regex.
First it checks if it has more than 2 forward slashes and gets
the pattern and flags. Then we finally check if it's a regex.
If it's not we just return it as a string.

Then the final array is returned as mix of strings and regexes.

batbattur commented:
deploy_block 🚛  on https://github.com/iFixit/ifixit/pull/47367 getting CR-ed
un_dev_block 🚛 fixed https://github.com/iFixit/css-gather/pull/9#discussion_r1165975217 in https://github.com/iFixit/css-gather/pull/9/commits/89f826b2ee03e83ac613e1d0e8d9cd24c9624e47

<hr>

Steps used to test:
1. Before excluding any elements from the critical css, confimed that `#mainBody` css exists.
![Screen Shot 2023-04-13 at 5 44 34 PM](https://user-images.githubusercontent.com/58952979/231913274-bacfd8a1-744a-4107-bea6-401376c8218a.png)
2. Then ran the run.rb script with `--exclude=/mAiNbOdy/gi` option:
`./run.rb https://www.ifixit.com/ --exclude=.skip-to-content --exclude=/mAiNbOdy/gi --exclude=/invalidregex > after ` 
3. Then verified that passing regex with `gi` flags works properly now. 
![Screen Shot 2023-04-13 at 5 44 54 PM](https://user-images.githubusercontent.com/58952979/231912862-7749e0b2-96ee-40c0-8280-89fa288553d0.png)|

QA 🚛

ardelato reviewed (approved):
CR ⚡

batbattur commented:
un_deploy_block 🚛 the pull on `iFixit/ifixit` side got reviewed
